gnome-session/gnome-login-check.1
.\" This manpage has been automatically generated by docbook2man 
.\" from a DocBook document.  This tool can be found at:
.\" <http://shell.ipoline.com/~elmert/comp/docbook2X/> 
.\" Please send any bug reports, improvements, comments, patches, 
.\" etc. to Steve Cheng <steve@ggi-project.org>.
.TH "GNOME-LOGIN-CHECK" "1" "21 oktober 2001" "" ""
.SH NAME
gnome-login-check \- Tool for gnome-session
.SH SYNOPSIS

\fBgnome-login-check\fR

.SH "DESCRIPTION"
.PP
\fBgnome-login-check\fR is needed by gnome-session to
do some checks during login.
.SH "OPTIONS"
.PP
No options. This program is called internally by gnome-session.
.SH "AUTHOR"
.PP
This manual page was written by Christian Marillat <marillat@debian.org> for
the Debian GNU/Linux system (but may be used by others).


NEWS
============
Version 51.0
============

- Fixed bug where oo7-daemon.service never starts by default
- Translation updates

=============
Version 51.rc
=============

- Switched to dbus for looking up logind graphical session
- Downgraded logout inhibitors from strong to weak, as a stop-gap fix for
  buggy inhibitor behavior
- Translation updates

===============
Version 51.beta
===============

- Added `SessionIsLocked` and `SessionClass` properties on D-Bus
- Updated mimeapps.list: removed Shaper, since it was split out of GTK into its
  own app
- Fixed regression in screen blanking / idle handling
- Translation updates

================
Version 51.alpha
================

- Enabled use of oo7-portal as the implementation of the Secret portal
- Updated mimeapps.list for removal of gcr3, and expansion of Loupe's supported
  mime types
- Applied various code cleanups from gobject linter
- Fixed a double-free bug
- Translation updates

============
Version 50.0
============

- No changes since 50.rc, just a version number bump and re-release!

=============
Version 50.rc
=============

- Hid availability of session save/restore support. This feature has been postponed to
  a future release of GNOME
- Fixed bug where inhibitors list doesn't update in end-session dialog
- Refactored CanShutdown() to more accurately represent the availability of
  the shutdown action, including API-incompatible changes that allow it to
  report inhibitor status, needs-authentication status, and so on
- Introduced CanReboot(), which enables the rest of the session to hide reboot
  buttons if rebooting isn't actually available
- Introduced CanSuspend() and Suspend(), for putting the system to sleep
- GNOME's logind sleep inhibitor was downgraded from a type='block' to a
  type='block-weak'. This restores the behavior of the inhibitor prior to
  systemd v257
- We now ignore sleep inhibitors when the user manually suspends the system. The
  user will likely need to authenticate before ignoring the inhibitors. Note that
  systemd will, by default, also automatically ignore sleep inhibitors when a
  laptop's lid is closed.
- Translation updates

===============
Version 50.beta
===============

- Updates to mimeapps.list: Incorporated latest changes to Font Viewer, Software,
  and Loupe
- Translation updates

================
Version 50.alpha
================

- Updated documentation to reflect the major changes in GNOME 49
- Fixed bug where gnome-session would fail to reap the processes of the apps
  it launched, leaving around zombie processes
- Fixed option parsing regression in gnome-session-inhibit
- Reworked logic to find DBUS_SESSION_BUS_ADDRESS, which should make gnome-session
  work better in non-systemd environments
- Removed version number from gettext POT name, which should make translations
  easier on services like launchpad.net
- Fixed bug with offline updates that prevented gnome-shell from converting a
  shutdown into a reboot
- Improved safety check that enforces only one active graphical session per user by
  checking for any running graphical session, not just another GNOME session
- Dropped support for running X11 sessions. Note that this dropped all display server
  specific unit files, since the generic gnome-session*.target units are now Wayland-only
- Implemented initial plumbing for session save/restore
- Reworked unit files to adapt to new changes where gnome-shell's unit file takes
  the gnome-shell mode as its argument. This makes it easier for kiosk sessions or
  other special/custom sessions to use gnome-shell modes
- Dropped unused dependencies: json-glib and gtk4
- Updates to mimeapps.list: Added Shaper (new GTK icon editor tool) and fixed bugs
  with GNOME Disk Utility
- Translation Updates

============
Version 49.0
============

- Translation updates
- Fixed build with musl libc
- Fixed some memory corruption bugs that led to gnome-session-init-worker
  crashing on logout
- Fixed a memory leak in gnome-session-init-worker
- Allowed Qt apps to use the Wayland text-input protocol, instead of forcing
  them to use the direct ibus integration plugin

=============
Version 49.rc
=============

- Disabled the org.freedesktop.ScreenSaver proxy service on the login screen
- Separated out the systemd-specific code from the gnome-session leader process
  into a dedicated executable. This makes it easier to integrate distributions
  that use alternative init systems with gnome-session (though note that this
  is still a usecase we don't officially support upstream)
- Translation updates

===============
Version 49.beta
===============

- gnome-session is no longer started via a shell script, and instead all of
  the necessary setup is done in C.
- gnome-session's builtin service manager has been completely removed. Service
  management will now be handled entirely by the operating system's native
  service manager (i.e. systemd). See the following blog post for more information:
  https://blogs.gnome.org/adrianvovk/2025/06/10/gnome-systemd-dependencies
- The "fail whale" dialog has been removed
- Various unused APIs and interfaces were removed, including:
	- The org.gnome.SessionManager.Client and o.g.SM.App D-Bus interfaces
	- Various D-Bus APIs related to the service manager
	- Various .desktop and .session file keys that were used only by the
	  builtin service manager
	- Fallback sessions, since there's no way to fall back anymore
- The code was significantly refactored
- gnome-session now provides a gnome-session-basic-services.target, that sessions
  like the GDM greeter and gnome-initial-setup can use to avoid listing all of
  the basic GNOME session services (like all the gnome-settings-daemon plugins)
- The --autostart command line option was removed, and instead kiosk sessions
  should set Kiosk=true in their session definition file.
- Updated GNOME's mimetype defaults to drop Evince (which was replaced with Papers)
- Translation updates

==================
Version 49.alpha.1

gnome-session/gnome-login-check.c
/* gnome-login-check.c - Do some checks during login

   Written by Owen Taylor <otaylor@redhat.com>
   Copyright (C) Red Hat

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
   02111-1307, USA.  */

#include <config.h>
#include <gnome.h>
#include <gdk/gdkx.h>
#include <stdio.h>
#include <netinet/in.h>
#include <sys/wait.h>
#include <errno.h>
#include <limits.h>
#include <pwd.h>
#include <sys/stat.h>

GtkWidget *dns_dialog = NULL;
GtkWidget *progress = NULL;
gint input_tag = FALSE;
gint timeout_tag = FALSE;

/* returns the hostname */
static gchar *
get_hostname (gboolean readable)
{
	static gboolean init = FALSE;
	static gchar *result = NULL;
	
	if (!init){
		char hostname[256];
		
		if (gethostname (hostname, sizeof (hostname)) == 0)
			result = g_strdup (hostname);
		
		init = TRUE;
	}
	
	if (result)
		return result;
	else
		return readable ? "(Unknown)" : NULL;
}

/* Timeout function to update the activity progress bar */
static gint
timeout (gpointer data)
{
	gint *progress_amount = data;
	float amount;
	
	if (!dns_dialog) {
		gchar *msg = g_strdup_printf (_("Looking up internet address for %s"), get_hostname(TRUE));
		
		dns_dialog = gnome_message_box_new (msg,
						    GNOME_MESSAGE_BOX_INFO,
						    _("Hide"),
						    NULL);
		g_free (msg);

		gnome_dialog_close_hides (GNOME_DIALOG (dns_dialog), TRUE);
		gtk_window_set_position (GTK_WINDOW (dns_dialog), GTK_WIN_POS_CENTER);
		gtk_window_set_title (GTK_WINDOW (dns_dialog), "Address Lookup");

		progress = gtk_progress_bar_new ();
		gtk_progress_set_activity_mode (GTK_PROGRESS (progress), TRUE);
		gtk_progress_bar_set_activity_blocks (GTK_PROGRESS_BAR (progress), 10);
		gtk_progress_bar_set_activity_step (GTK_PROGRESS_BAR (progress), 5);
		
		gtk_widget_show (progress);
		gtk_box_pack_start(GTK_BOX(GNOME_DIALOG(dns_dialog)->vbox),
				   progress,
				   FALSE, TRUE, 0);

		gtk_widget_show_all (dns_dialog);
	}

	amount = *progress_amount % 18;
	if (amount > 9)
		amount = 18 - amount;

	amount = amount / 10;

	gtk_progress_set_percentage (GTK_PROGRESS (progress), amount);
	*progress_amount = *progress_amount + 1;
	
	return TRUE;
}

static gboolean
startup_timeout (gpointer data)
{
	timeout_tag = gtk_timeout_add (50, timeout, data);
	return FALSE;
}

/* Called when dns-helper reports its result */
static void
input_callback (gpointer data, gint fd, GdkInputCondition condition)
{
	gint count;
	struct in_addr ip_addr;
	gboolean *success = data;

	count = read (fd, &ip_addr, sizeof(ip_addr));

	if (count != sizeof(ip_addr)) {
		g_print ("here\n");
		*success = FALSE;
	} else {
		*success = (ip_addr.s_addr != 0);
	}

	wait(NULL);

	gdk_input_remove (input_tag);
	gtk_timeout_remove (timeout_tag);

	if (dns_dialog) {
		gtk_widget_destroy (dns_dialog);
		dns_dialog = NULL;
		progress = NULL;
	}
	
	gtk_main_quit ();
}

/* Check if a DNS lookup on `hostname` can be done */
static gboolean
check_for_dns (void)
{
	gchar *hostname;

	gboolean success;
	gint progress_amount = 0;

	hostname = get_hostname(FALSE);
	
	if (!hostname)
		return FALSE;
	else {
		gint pid;
		int in_fd[2];
		int out_fd[2];
		
		if (pipe(in_fd))
			return FALSE;
		if (pipe(out_fd))
			return FALSE;
		
		if (!(pid = fork())) { /* Child */
			close (in_fd[1]);
			close (out_fd[0]);
			
			dup2 (in_fd[0], STDIN_FILENO);
			dup2 (out_fd[1], STDOUT_FILENO);
			
			execlp ("dns-helper", "dns-helper", NULL);

			/* Should never be reached */
			_exit (0);

		} else if (pid > 0) { /* Parent */
			write (in_fd[1], hostname, strlen(hostname));
			close (in_fd[1]);
			close (in_fd[0]);
			
			close (out_fd[1]);
			
			/* GDK_INPUT_EXCEPTION is a temporary hack here */
			input_tag = gdk_input_add (out_fd[0], GDK_INPUT_READ | GDK_INPUT_EXCEPTION,
						   input_callback, &success);
			timeout_tag = gtk_timeout_add (1000, startup_timeout, &progress_amount);

			gtk_main ();
			
			return success;
		  
		} else
			return FALSE;
	}
}

static gboolean
check_orbit_dir(void)
{
  char buf[PATH_MAX];
  struct stat sbuf;

  sprintf(buf, "/tmp/orbit-%s", g_get_user_name());
  if(stat(buf, &sbuf))
    return TRUE; /* Doesn't exist - things are fine */

  if(sbuf.st_uid != getuid()
     && sbuf.st_uid != geteuid())
    return FALSE;

  return TRUE;
}

int
main (int argc, char **argv)
{
        GdkModifierType state;
	GtkWidget *dialog;
	GtkWidget *msgbox;
	GtkWidget *hbox;
	GtkWidget *vbox;
	GtkWidget *session_toggle;
	GtkWidget *settings_toggle;
	GtkWidget *frame;
	GtkWidget *pixmap = NULL;
	gchar *msg, *s;
  
	gnome_init ("gnome-login-check", "0.1", argc, argv);

	gdk_window_get_pointer (GDK_ROOT_PARENT(), NULL, NULL, &state);

	if ((state & (GDK_CONTROL_MASK|GDK_SHIFT_MASK)) == (GDK_CONTROL_MASK|GDK_SHIFT_MASK)) {
		dialog = gnome_dialog_new (_("GNOME Login"), _("Login"), NULL);
		gtk_window_set_position (GTK_WINDOW (dialog), GTK_WIN_POS_CENTER);

		hbox = gtk_hbox_new (FALSE, GNOME_PAD);
		gtk_box_pack_start (GTK_BOX (GNOME_DIALOG (dialog)->vbox),
				    hbox, FALSE, FALSE, 0);

		s = gnome_unconditional_pixmap_file("gnome-logo-large.png");
		if (s) {
                        pixmap = gnome_pixmap_new_from_file(s);
                        g_free(s);
			
			frame = gtk_frame_new (NULL);
			gtk_container_set_border_width (GTK_CONTAINER (frame),
							GNOME_PAD_SMALL);
			gtk_frame_set_shadow_type (GTK_FRAME (frame),
						   GTK_SHADOW_IN);
			gtk_container_add (GTK_CONTAINER (frame), pixmap);
			
			gtk_box_pack_start (GTK_BOX (hbox),
					    frame, FALSE, FALSE, 0);
                }

		vbox = gtk_vbox_new (FALSE, GNOME_PAD_SMALL);
		gtk_box_pack_start (GTK_BOX (hbox),
				    vbox, FALSE, FALSE, 0);

		session_toggle = gtk_check_button_new_with_label (_("Start with default programs"));
		gtk_box_pack_start (GTK_BOX (vbox), session_toggle,
				    FALSE, FALSE, 0);

		settings_toggle = gtk_check_button_new_with_label (_("Reset all user settings"));
		gtk_box_pack_start (GTK_BOX (vbox), settings_toggle,
				    FALSE, FALSE, 0);
	    
		gnome_dialog_close_hides (GNOME_DIALOG (dialog), TRUE);

		while (1) {
		        gtk_widget_show_all (dialog);
			gnome_dialog_run (GNOME_DIALOG (dialog));

			if (GTK_TOGGLE_BUTTON (settings_toggle)->active) {
				msg = g_strdup_printf (_("Really reset all GNOME user settings for %s?"), g_get_user_name());
				msgbox = gnome_message_box_new (msg,
								GNOME_MESSAGE_BOX_QUESTION,
								GNOME_STOCK_BUTTON_OK,
								GNOME_STOCK_BUTTON_CANCEL,
								NULL);
				gtk_window_set_position (GTK_WINDOW (msgbox),
							 GTK_WIN_POS_CENTER);

				if (gnome_dialog_run (GNOME_DIALOG (msgbox)) != 0)
					continue;
			}

			break;
		}
		
		if (GTK_TOGGLE_BUTTON (settings_toggle)->active) {
			gchar *filename = gnome_util_home_file ("");
			gchar *command = g_strconcat ("rm -rf ", filename, NULL);
			system (command);
			g_print("%s\n", command);
		} else if (GTK_TOGGLE_BUTTON (session_toggle)->active) {
			gchar *filename = gnome_util_home_file ("session");
			gchar *command = g_strconcat ("rm ", filename, NULL);
			system (command);
		}
	}

	if (!check_orbit_dir ()) {
		GtkWidget *tmp_msgbox;
		gchar *tmp_msg;
		tmp_msg = g_strdup_printf (
			_("The directory /tmp/orbit-%s is not owned\nby the current user, %s.\n"
			  "Please correct the ownership of this directory."),
			g_get_user_name(), g_get_user_name());
		
		tmp_msgbox = gnome_message_box_new (
			tmp_msg,
			GNOME_MESSAGE_BOX_WARNING,
			_("Try again"),
			_("Continue"),
			NULL);
	  
		gtk_window_set_position (GTK_WINDOW (tmp_msgbox), GTK_WIN_POS_CENTER);
		gnome_dialog_close_hides (GNOME_DIALOG (tmp_msgbox), TRUE);
		
		while ((gnome_dialog_run (GNOME_DIALOG (tmp_msgbox)) == 0) &&
		       !check_orbit_dir ())
			/* Nothing */;
	}
	
	if (!check_for_dns ()) {
		GtkWidget *tmp_msgbox;
		gchar *tmp_msg;
		
		tmp_msg = g_strdup_printf (
		  _("Could not look up internet address for %s.\n"
		    "This will prevent GNOME from operating correctly.\n"
		    "It may be possible to correct the problem by adding\n"
		    "%s to the file /etc/hosts."),
		  get_hostname(TRUE), get_hostname(TRUE));
		
		tmp_msgbox = gnome_message_box_new (tmp_msg,
						GNOME_MESSAGE_BOX_WARNING,
						_("Try again"),
						_("Continue"),
						NULL);

		gtk_window_set_position (GTK_WINDOW (tmp_msgbox), GTK_WIN_POS_CENTER);
		gnome_dialog_close_hides (GNOME_DIALOG (tmp_msgbox), TRUE);

		while ((gnome_dialog_run (GNOME_DIALOG (tmp_msgbox)) == 0) &&
		       !check_for_dns ())
			/* Nothing */;
			
	}

	return 0;
}
